codelabs/theme_extension/theme_extension.md

```diff
@@ -0,0 +1,189 @@
+
+[Download zip](https://github.com/google/blockly-samples/archive/master.zip)
+
+or by cloning this git repo:
+
+```bash
+git clone https://github.com/google/blockly-samples.git
+```
+
+If you downloaded the source as a zip, unpacking it should give you a root folder named `blockly-samples-master`.
+
+The relevant files are in `examples/theme-extension-codelab`. There are two versions of the app:
+- `starter-code/`: The starter code that you'll build upon in this codelab.
+- `complete-code/`: The code after completing the codelab, in case you get lost or want to compare to your version.
+
+Each folder contains:
+- `index.js` - The codelab's logic. To start, it just injects a simple workspace.
+- `index.html` - A web page containing a simple blockly workspace.
+
+To run the code, simple open `starter-code/index.html` in a browser. You should see a Blockly workspace with a flyout.
+
+![A web page with the text "Theme Extension Codelab" and a Blockly workspace.](starter_workspace.png)
+
+## Workspace Theme
+
+In this section you will create a very basic `Blockly.Themes.Halloween`, then inject it to display in the workspace.
+
+### Themes
+
+Themes are a way to customize the look and feel of Blockly. Currently, we support customizing block colours, category colours and certain components through the Themes class
+
+A theme can be created using the constructor or by using defineTheme. Using defineTheme makes it easy to extend a pre existing theme and set all values with a single object.
+
+Add the below code in `index.js` right before function start()
+```js
+Blockly.Themes.Halloween = Blockly.Theme.defineTheme('halloween', {
+  'base': Blockly.Themes.Classic,
+});
+```
+
+And then add a line to inject this theme in function start() in `index.js`
+```js
+function start() {
+    // Create main workspace.
+    workspace = Blockly.inject('blocklyDiv',
+      {
+        toolbox: toolboxCategories,
+        theme: Blockly.Themes.Halloween,
+      });
+ } 
+```
+
+Note: At this point, we have just created a new theme. But it does not have any customizations. We will do that next.
+
+## Customize Components 
+
+Within the Halloween theme definition, you can customize the colours of multiple components:
+
+```js
+Blockly.Themes.Halloween = Blockly.Theme.defineTheme('halloween', {
+  'base': Blockly.Themes.Classic,
+  'componentStyles': {
+    'workspaceBackgroundColour': '#ff7518',
+    'toolboxBackgroundColour': '#F9C10E',
+    'toolboxForegroundColour': '#fff',
+    'flyoutBackgroundColour': '#252526',
+    'flyoutForegroundColour': '#ccc',
+    'flyoutOpacity': 1,
+    'scrollbarColour': '#ff0000',
+    'insertionMarkerColour': '#fff',
+    'insertionMarkerOpacity': 0.3,
+    'scrollbarOpacity': 0.4,
+    'cursorColour': '#d0d0d0',
+    'blackBackground': '#333'
+  }
+});
+```
+
+### Test it
+
+Reload your web page. You should see a themed workspace!
+
+
+![Fun with colours.](theme_components_workspace.png)
+
+## Customize Category Styles 
```

Comment:
I think there are two things going on here. We are setting the category styles and the block styles. I think it would be better to break these up into a "Customize Category Styles" and "Customize Block Styles" sections.